bluefog/common/topology_util.py

```diff
@@ -351,5 +425,160 @@ def GetDynamicSendRecvRanks(topo: nx.DiGraph, self_rank: int) -> Iterator[Tuple[
+
+    Args:
+        world_size (int): the size of all nodes; world_size = num_machines * nodes_per_machine
+        local_size (int): number of nodes in each machine
+        self_rank (int): The self rank.
+
+    Yields:
+        Iterator[Tuple[List[int], List[int]]]: send_ranks, recv_ranks.
+
+    Example:
+
+        >>> from bluefog.common import topology_util
+        >>> world_size, local_size = bf.size(), bf.local_size()
+        >>> gen = topology_util.GetInnerOuterRingDynamicSendRecvRanks(world_size, local_size, 0)
+        >>> for _ in range(10):
+        >>>     print(next(gen))
+    """
+    num_machines = world_size//local_size
+    nodes_per_machine = local_size
+    assert world_size % local_size == 0, "It should be used under homogeneous environment only."
+    assert nodes_per_machine > 2, "We do not support the case nodes per machine < 2 yet."
+
+    index = 0
+    while True:
+        machine_id = self_rank // nodes_per_machine
+        local_rank_id = self_rank % nodes_per_machine
+        local_rank_to_go_outside_id = index % nodes_per_machine
+
+        recv_ranks = []
+        if local_rank_to_go_outside_id == local_rank_id:
+            # find send_rank
+            target_machine_id = (machine_id + 1) % num_machines
+            target_rank_id = target_machine_id * nodes_per_machine + local_rank_id
+            send_rank = target_rank_id
+
+            # find recv_rank
+            source_machine_id = (machine_id - 1) % num_machines
+            source_rank_id = source_machine_id * nodes_per_machine + local_rank_id
+            recv_ranks.append(source_rank_id)
+
+        else:
+            if nodes_per_machine == 2:
+                # Do not send. But our neighbor_allreduce haven't supported this yet.
+                raise ValueError(
+                    "Unfortunately, nodes_per_machine is 2 case is not supported yet.")
+            else:
+                # find send_rank
+                target_local_rank_id = (local_rank_id + 1) % nodes_per_machine
+                if target_local_rank_id == local_rank_to_go_outside_id:
+                    target_local_rank_id = (
+                        target_local_rank_id + 1) % nodes_per_machine
+                target_rank_id = target_local_rank_id + machine_id * nodes_per_machine
+                send_rank = target_rank_id
+
+                # find recv_rank
+                source_local_rank_id = (local_rank_id - 1) % nodes_per_machine
+                if source_local_rank_id == local_rank_to_go_outside_id:
+                    source_local_rank_id = (
+                        source_local_rank_id - 1) % nodes_per_machine
+                source_rank_id = source_local_rank_id + machine_id * nodes_per_machine
+                recv_ranks.append(source_rank_id)
+
+        yield [send_rank], recv_ranks
+        index += 1
+
+
+def GetInnerOuterExp2DynamicSendRecvRanks(
+        world_size: int, local_size: int, self_rank: int
+    ) -> Iterator[Tuple[List[int], List[int]]]:
```

Comment:
Same here. The logic in this function seems inner machine is a exp2 topology, which is not the case for previous topology generation. Is there any inconsistency here?

Reply:
Discussed offline. Mainly due to the dynamic graph is the subgraph of the static graph requirement.